plugins/module_utils/foreman_helper.py

```diff
@@ -285,14 +287,108 @@ def ensure_resource(self, resource, entity_dict, entity, state=None, name_map=No
             self.fail_json(msg='Not a valid state: {}'.format(state))
         return changed, changed_entity
 
+    @_exception2fail_json('Failed to ensure entity state: %s')
+    def ensure_entity(self, resource, desired_entity, current_entity, params=None, state=None, entity_spec=None):
+        """Ensure that a given entity has a certain state
+
+            Parameters:
+                resource (string): Plural name of the api resource to manipulate
+                desired_entity (dict): Desired properties of the entity
+                current_entity (dict, None): Current properties of the entity or None if nonexistent
+                params (dict): Lookup parameters (i.e. parent_id for nested entities) (optional)
+                state (dict): Desired state of the entity (optionally taken from the module)
+                entity_spec (dict): Description of the entity structure (optionally taken from module)
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        if state is None:
+            state = self.state
+        if entity_spec is None:
+            entity_spec = self.entity_spec
+
+        changed = False
+
+        if state == 'present_with_defaults':
+            if current_entity is None:
+                changed, current_entity = self._create_entity(resource, desired_entity, params, entity_spec)
+        elif state == 'present':
+            if current_entity is None:
+                changed, current_entity = self._create_entity(resource, desired_entity, params, entity_spec)
+            else:
+                changed, current_entity = self._update_entity(resource, desired_entity, current_entity, params, entity_spec)
+        elif state == 'absent':
+            if current_entity is not None:
+                changed, current_entity = self._delete_entity(resource, current_entity, params)
+        else:
+            self.fail_json(msg='Not a valid state: {}'.format(state))
+        return changed, current_entity
```

Comment:
It kinda feels weird that you return `current_entity` and not `updated_entity`. Don't think it hurts at all, just found it weird when reading that you technically overwrite `current`.

Reply:
I can make it copy_on_write. If it helps readabilitiy.

Reply:
I don't think you have to, but you certainly can :)